enzyme/Enzyme/EnzymeLogic.cpp

```diff
@@ -4815,6 +4816,539 @@ Function *EnzymeLogic::CreateForwardDiff(
+        auto nres = cast<BinaryOperator>(B.CreateFSub(newLHS, newRHS));
+        nres->takeName(newI);
+        nres->copyIRFlags(newI);
+        newI->replaceAllUsesWith(expand(B, nres));
+        newI->eraseFromParent();
+      }
+        return;
+      case BinaryOperator::FDiv: {
+        auto nres = cast<BinaryOperator>(B.CreateFDiv(newLHS, newRHS));
+        nres->takeName(newI);
+        nres->copyIRFlags(newI);
+        newI->replaceAllUsesWith(expand(B, nres));
+        newI->eraseFromParent();
+      }
+        return;
+      case BinaryOperator::FRem: {
+        auto nres = cast<BinaryOperator>(B.CreateFRem(newLHS, newRHS));
+        nres->takeName(newI);
+        nres->copyIRFlags(newI);
+        newI->replaceAllUsesWith(expand(B, nres));
+        newI->eraseFromParent();
+      }
+        return;
+      }
+    }
+    todo(BO);
+    return;
+  }
+  void visitMemSetInst(llvm::MemSetInst &MS) { visitMemSetCommon(MS); }
+  void visitMemSetCommon(llvm::CallInst &MS) { return; }
+  void visitMemTransferInst(llvm::MemTransferInst &MTI) {
+    using namespace llvm;
+    Value *isVolatile = getNewFromOriginal(MTI.getOperand(3));
+    auto srcAlign = MTI.getSourceAlign();
+    auto dstAlign = MTI.getDestAlign();
+    visitMemTransferCommon(MTI.getIntrinsicID(), srcAlign, dstAlign, MTI,
+                           MTI.getOperand(0), MTI.getOperand(1),
+                           getNewFromOriginal(MTI.getOperand(2)), isVolatile);
+  }
+  void visitMemTransferCommon(llvm::Intrinsic::ID ID, llvm::MaybeAlign srcAlign,
+                              llvm::MaybeAlign dstAlign, llvm::CallInst &MTI,
+                              llvm::Value *orig_dst, llvm::Value *orig_src,
+                              llvm::Value *new_size, llvm::Value *isVolatile) {
+    return;
+  }
+  void visitFenceInst(llvm::FenceInst &FI) { return; }
+  void visitIntrinsicInst(llvm::IntrinsicInst &II) {
+    SmallVector<Value *, 2> orig_ops(II.arg_size());
+    for (unsigned i = 0; i < II.arg_size(); ++i)
+      orig_ops[i] = II.getOperand(i);
+    if (handleAdjointForIntrinsic(II.getIntrinsicID(), II, orig_ops))
+      return;
+
+    bool hasFromType = false;
+    auto newI = cast<llvm::IntrinsicInst>(getNewFromOriginal(&II));
+    IRBuilder<> B(newI);
+    SmallVector<Value *, 2> new_ops(II.arg_size());
+    for (unsigned i = 0; i < II.arg_size(); ++i) {
+      if (orig_ops[i]->getType() == getFromType()) {
+        new_ops[i] = truncate(B, getNewFromOriginal(orig_ops[i]));
+        hasFromType = true;
+      } else {
+        new_ops[i] = getNewFromOriginal(orig_ops[i]);
+      }
+    }
+    Type *retTy = II.getType();
+    if (II.getType() == getFromType()) {
+      hasFromType = true;
+      retTy = getToType();
+    }
+
+    if (!hasFromType)
+      return;
+
+    // TODO check that the intrinsic is overloaded
+
+    CallInst *intr;
+#if LLVM_VERSION_MAJOR >= 16
+    Value *nres = intr = B.CreateIntrinsic(retTy, II.getIntrinsicID(), new_ops,
```

Comment:
can we just make CreateIntrinsic(B, retTy, intrId, ops, ...) as a fn in utils.h that contains this?